tests/spiped_nc/graceful_shutdown.c

```diff
@@ -0,0 +1,122 @@
+#include <signal.h>
+#include <stdlib.h>
+
+#include "events.h"
+#include "warnp.h"
+
+#include "graceful_shutdown.h"
+
+struct graceful_shutdown_cookie {
+	int (* begin_shutdown)(void *);
+	void (* sighandler_sigterm_orig)(int);
+	void * caller_cookie;
+	void * timer_cookie;
+};
+
+static volatile sig_atomic_t should_shutdown = 0;
+
+/* Signal handler for SIGTERM to perform a graceful shutdown. */
+static void
+graceful_shutdown_handler(int signo)
+{
+
+	(void)signo; /* UNUSED */
+	should_shutdown = 1;
+}
+
+/* Requests a graceful shutdown of the caller via the cookie info. */
+static int
+graceful_shutdown(void * cookie)
+{
+	struct graceful_shutdown_cookie * G = cookie;
+
+	/* This timer has expired. */
+	G->timer_cookie = NULL;
+
+	/* Use the callback function, or schedule another check in 1 second. */
+	if (should_shutdown) {
+		if (G->begin_shutdown(G->caller_cookie) != 0) {
+			warnp("Failed to begin shutdown");
+			goto err0;
+		}
+	} else {
+		if ((G->timer_cookie = events_timer_register_double(
+		    graceful_shutdown, G, 1.0)) == NULL) {
+			warnp("Failed to register the graceful shutdown timer");
+			goto err0;
+		}
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * graceful_shutdown_register(callback, caller_cookie):
+ * Initializes an event which checks for SIGTERM every 1.0 seconds; if
+ * detected, it calls the ${callback} function and passes it the ${cookie}.
+ */
+void *
+graceful_shutdown_register(int (* begin_shutdown)(void *),
```

Comment:
Is it your intention that more than one graceful shutdown callback might be registered at once?

If yes, you need to make sure you restore signal handlers in the right order.

If no, why bother returning a cookie when you could just keep all the parameters as static variables inside this file?